zeppelin-zengine/src/main/java/org/apache/zeppelin/interpreter/RunNotebookInterpreter.java

```diff
@@ -0,0 +1,294 @@
+    } catch (IOException e) {
+      return new InterpreterResult(InterpreterResult.Code.ERROR, ExceptionUtils.getStackTrace(e));
+    } finally {
+      if (lastParagraphId != null) {
+        try {
+          String finalLastParagraphId = lastParagraphId;
+          notebook.processNote(context.getNoteId(), note -> {
+            note.removeParagraph("anonymous", finalLastParagraphId);
+            return null;
+          });
+        } catch (IOException e) {
+          LOGGER.error(String.format("Fail to remove paragraph: %s", lastParagraphId), e);
+        }
+      }
+    }
+  }
+
+  private InterpreterResult runNote(InterpreterContext context,
+                                    Paragraph lastParagraph,
+                                    String noteId,
+                                    Map<String, String> params,
+                                    int outputStartIndex) {
+    List<InterpreterResultMessage> allResultMessages = new ArrayList<>();
+    try {
+      // copy execution note
+      boolean isSuccess = notebook.processNote(noteId, note -> {
+        if (note == null) {
+          allResultMessages.add(new InterpreterResultMessage(InterpreterResult.Type.TEXT,
+                  String.format("Note `%s` doesn't exist", noteId)));
+          return false;
+        }
+        int outputIndex = outputStartIndex;
+        for (Paragraph p : note.getParagraphs()) {
+          if (StringUtils.isBlank(p.getScriptText())) {
+            // skip empty paragraph
+            continue;
+          }
+          lastParagraph.setTitle(p.getTitle());
+          lastParagraph.setText(p.getText());
+          lastParagraph.setConfig(p.getConfig());
+          lastParagraph.settings.setParams(new HashMap<>(p.settings.getParams()));
+          lastParagraph.settings.getParams().putAll(params);
+          lastParagraph.settings.setForms(new HashMap<>(p.settings.getForms()));
+          lastParagraph.run();
+          InterpreterResult result = lastParagraph.getReturn();
+          for (InterpreterResultMessage resultMessage : result.message()) {
+            interpreterProcessListener.onOutputUpdated(context.getNoteId(), context.getParagraphId(), outputIndex ++,
+                    resultMessage.getType(), resultMessage.getData());
+            allResultMessages.add(resultMessage);
+          }
+          if (result.code != InterpreterResult.Code.SUCCESS) {
+            return false;
+          }
+        }
+        return true;
+      });
+
+      if (isSuccess) {
+        return new InterpreterResult(InterpreterResult.Code.SUCCESS, allResultMessages);
+      } else {
+        return new InterpreterResult(InterpreterResult.Code.ERROR, allResultMessages);
+      }
+    } catch (IOException e) {
+      return new InterpreterResult(InterpreterResult.Code.ERROR, ExceptionUtils.getStackTrace(e));
+    }
+  }
+
+  @VisibleForTesting
+  public static Map<String, String> parseParams(String text) throws InvalidParameterFormatException {
```

Comment:
Just curious. Why don't we use regex to handle it? It also handle some exceptional cases but I wonder if we need to handle that case or not.